### ISSUE TITLE
Check if deprecation comment is in a seperate paragraph.

### ISSUE THREAD

**Issue by jvanz** [go-critic/go-critic#1453]
The Go documentation states that a deprecated comment should be in a separate paragraph:

> To signal that an identifier should not be used, add a paragraph to its doc comment that begins with Deprecated: followed by some information about the deprecation, and a recommendation on what to use instead, if applicable. The paragraph does not have to be the last paragraph in the doc comment.

However, the go-critic's `deprecatedComment` checker is not detecting this issue, which prevents users from identifying old code that will be removed in the near future and preparing to make necessary changes. Therefore, I would like to request that you implement this check as well.

**Comment by quasilyte** [comment]
`deprecatedComment` is a good place to include this check in my opinion

**Comment by ccoVeille** [comment]
I'm facing a similar issue by working on a revive rule for something that also involves the detection of comments flagged `Deprecated:`

Here is somehow the target

```go
// Foo is foo
//
// Deprecated: Lorem Ipsum
func Foo() {}
```

Here are the diversity of "format" I have found. So I'm sharing them with you, guys.

This one is the one you are talking about
```go
// Foo is foo
// Deprecated: 
func Foo() {}
```

This one is frequent, no paragraph
```go
// Deprecated: Lorem Ipsum
// Foo is foo
func Foo() {}
```

This one, I'm unsure if it's a syntax that should be considered as valid, or if should be reported. 

```go
// Deprecated: Lorem Ipsum
//
// Foo is foo
func Foo() {}
```

Please note I have also faced these, where people misunderstood the logic of paragraph

```go
// Deprecated: Lorem Ipsum

// Foo is foo
func Foo() {}
```

and this

```go
// Foo is foo

// Deprecated: Lorem Ipsum
func Foo() {}
```

Please note, I didn't check them against what deprecatedComment or gocritic already report

I just wanted to share with you what I knew to exist as it may help you.

**Comment by ccoVeille** [comment]
My problems in revive are over these.

My problem are with multiline syntax like these

```go
// Foo is foo
// and bar
// because blah blah
// Deprecated: the reason
// why it's deprecated is
// because blah blab
func Foo() {}
```

The logic would be to have this, I think

```go
// Foo is foo
// and bar
// because blah blah
//
// Deprecated: the reason
// why it's deprecated is
// because blah blab
func Foo() {}
```

But

Sometimes the code is a variation of previous alternatives

```go
// Deprecated: the reason
// why it's deprecated is
// because blah blab
// Foo is foo
// and bar
// because blah blah
func Foo() {}
```

Or

```go
// Deprecated: the reason
// why it's deprecated is
// because blah blab
func Foo() {}
```

So now you know, you may say let's keep rules simple. But having these format and variations in mind may help you

**Comment by quasilyte** [comment]
@ccoVeille those are good points, thank you for sharing.

I guess for the sake of simplicity, we could use this as an approximation:

* Scan until `Deprecated:` prefix is found, check for empty lines
* If there is a `Deprecated:` prefix, enter the "deprecated block" state
* If there is no empty line and it's not a beginning of a comment (its first line), emit a warning

And then we can't disambiguate whether a deprecation comment block has irrelevant stuff in it or not. It seem to require a lot of heuristics. In other words, we can't tell if a deprecation comment has ended or not:

```
// Foo is foo,
// bar is bar,
// Go is good
//
// Deprecated: please use X instead
// and some other irrelevant text that should have
// had an empty line separating it from the deprecation comment
```

**Comment by ccoVeille** [comment]
I'm fine with your analysis.

Maybe checking at the go-tools lib is also interesting.

https://github.com/dominikh/go-tools/blob/0f2f29196d9a138e7beeed4e26e890cf9273f63e/analysis/facts/deprecated/deprecated.go#L13

**Comment by ccoVeille** [comment]
Please note the detection seems to be:
- what you suggested: anything after Deprecated: is part deprecated
- the section separator is `\n\nDeprecated: ` (with a trailing space)

So yes, the paragraph idea is important, and then it means there should be something before the deprecated paragraph, so a another paragraph must be present before.

So anything that differs should be reported

**Comment by cristaloleg** [comment]
```
// Deprecated: Lorem Ipsum

// Foo is foo
func Foo() {}
```

Lonely standing `// Deprecated: ` for me looks like something we can report. Maybe as a separate check, my 2c.

**Comment by mx-psi** [comment]
Filed #1470 for this (pretty similar to #1430 which I am not sure why it was closed)

**Comment by ccoVeille** [comment]
For record, there is a backlog, but inactive, issue about Deprecated paragraph vs the rest

- https://github.com/golang/go/issues/38743

if anyone is interested you can follow the discussion
